### PR TITLE
[bugfix](RowsetIterator) use valid stats when creating segment iterator

### PR DESCRIPTION
**Pull request by ByteYue**
## Proposed changes
The former logic would directly use read_context as segment iterator's _context `_context = read_context;`. But if it's invoked by schema change or compaction, it would pass one read_context whose stats is nullptr like the following picture.
image

Thus the following file cache statistics assignation would get one invalid address `_read_options.io_ctx.file_cache_stats = &read_context->stats->file_cache_stats;` which would be the offset for file_cache_stats in stats.

This pr would make the read_context->stats always valid to prevent such coredump.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by ByteYue** [comment]
run buildall

File: be/src/olap/rowset/beta_rowset_reader.cpp [hunk 1]
```diff
@@ -80,11 +80,19 @@ Status BetaRowsetReader::get_segment_iterators(RowsetReaderContext* read_context
...
+    // However, for compaction or schema change workloads,
+    // the read_context passed to the function will have null statistics,
+    // and in such cases we will try to use the beta rowset reader's own statistics.
     if (_context->stats != nullptr) {
-        // schema change/compaction should use owned_stats
-        // When doing schema change/compaction,
-        // only statistics of this RowsetReader is necessary.
         _stats = _context->stats;
+    } else {
+        _context->stats = _stats;
```
Comment (Tanya-W) [inline be/src/olap/rowset/beta_rowset_reader.cpp:95]: at line 211 should also be changed?
`_read_options.io_ctx.file_cache_stats = &read_context->stats->file_cache_stats;` to `_read_options.io_ctx.file_cache_stats = &_context->stats->file_cache_stats;`

Comment (ByteYue) [reply be/src/olap/rowset/beta_rowset_reader.cpp:95]: The member variable `_context` is already assigned at line 82 `_context = read_context;`. So the following read_context and _context are the same.

**Review by Tanya-W — APPROVED** [review]
LGTM

**Comment by ByteYue** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM

**Pull request merged**